docs/fundamentals/custom-resource-urls.md

```diff
@@ -0,0 +1,79 @@
+
+## Default endpoint behavior
+
+By default, as described in the [Networking inner loop](networking-overview.md#networking-in-the-inner-loop) article, .NET Aspire relies on existing configurations such as Kestrel or launch profiles to determine the host and port of a resource for a configured endpoint.
+
+Likewise, you can explicitly expose endpoints using the <xref:Aspire.Hosting.ResourceBuilderExtensions.WithEndpoint*> API. This API allows you to specify the host and port for a resource, which is then used to create the default URL for that resource. The default URL is typically in the format `http://<host>:<port>` or `https://<host>:<port>`, depending on the protocol used. To omit the host port, use one of the following methods:
+
+- <xref:Aspire.Hosting.ResourceBuilderExtensions.WithHttpEndpoint*>
+- <xref:Aspire.Hosting.ResourceBuilderExtensions.WithHttpsEndpoint*>
+
+For more information, see [Endpoint extension methods](networking-overview.md#endpoint-extension-methods).
+
+## Supported resource types
+
+Currently, custom resource URLs are supported for the following resource types:
+
+- <xref:Aspire.Hosting.ApplicationModel.ContainerResource>
+- <xref:Aspire.Hosting.ApplicationModel.ExecutableResource>
+- <xref:Aspire.Hosting.ApplicationModel.ProjectResource>
+
+## Customize resource URLs
+
+Use the appropriate `WithUrl` overload, `WithUrls`, or `WithUrlForEndpoint` APIs on any supported resource builder to define custom URLs for a resource. The following example demonstrates how to set a custom URL for a project resource:
+
+:::code source="snippets/custom-urls/AspireApp.AppHost/Program.WithUrl.cs" id="withurl":::
+
+> [!TIP]
+> There's an overload that accepts a `string` allowing you to pass any URL. This is useful for defining custom URLs that aren't directly related to the resource's endpoint.
+
+The preceding code assigns a project reference to the `api` variable, which is then used to create a custom URL for the `Admin Portal` route. The `WithUrl` method takes a <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> and a display name as parameters. The resulting URL is available in the dashboard as shown in the following screenshot:
+
+:::image type="content" source="dashboard/media/custom-urls/custom-url-admin-portal.png" alt-text=".NET Aspire dashboard custom Admin Portal URL." lightbox="dashboard/media/custom-urls/custom-url-admin-portal.png":::
+
+### Customize endpoint URL
+
+<!-- TODO: Add xref to WithUrlForEndpoint when available -->
+
+To expose specific endpoints like Scalar or Swagger in the dashboard, use the `WithUrlForEndpoint` method. The following example demonstrates how to customize the URL for a resource endpoint:
```

Comment:
This phrasing makes it sound like Scalar and Swagger are endpoints, but they're not. The endpoint is the scheme/host/port declaration and things like Scalar and Swagger are just services that a given resource might be serving from a URL based at a declared endpoint. It's important to call out that the `WithUrlForEndpoint` will **update** the first URL that is associated with the given endpoint. If you want to **add** a separate URL (even for the same endpoint) you need to call the `WithUrl` overload that takes a `ReferenceExpression` or interpolated string, or call `WithUrls` and add the URL to the `Urls` list on the context.

Reply:
Fixed